proposals/csharp-8.0/async-streams.md

```diff
@@ -187,52 +187,78 @@ await foreach (var i in enumerable)
+  - Otherwise, if there is more than one such type `T`, then an error is produced and no further steps are taken.
+- Otherwise, an error is produced and no further steps are taken.
 
-This code:
+The above steps, if successful, unambiguously produce a collection type `C`, enumerator type `E` and iteration type `T`.
 
 ```csharp
-var enumerable = ...;
-await foreach (T item in enumerable)
-{
-   ...
-}
+await foreach (V v in x) «embedded_statement»
 ```
 
-is translated to the equivalent of:
+is then expanded to:
 
 ```csharp
-var enumerable = ...;
-var enumerator = enumerable.GetAsyncEnumerator();
-try
 {
-    while (await enumerator.MoveNextAsync())
-    {
-       T item = enumerator.Current;
-       ...
+    E e = ((C)(x)).GetAsyncEnumerator();
+    try {
+        while (await e.MoveNextAsync()) {
+            V v = (V)(T)e.Current;
+            «embedded_statement»
+        }
+    }
+    finally {
+        ... // Dispose e
     }
-}
-finally
-{
-    await enumerator.DisposeAsync(); // omitted, along with the try/finally, if the enumerator doesn't expose DisposeAsync
 }
 ```
 
-If the iterated type doesn't expose the right pattern, the interfaces will be used.
+The body of the `finally` block is constructed according to the following steps:
+- If the type `E` has an appropriate `DisposeAsync` method:
+  - Perform member lookup on the type `E` with identifier `DisposeAsync` and no type arguments. If the member lookup does not produce a match, or it produces an ambiguity, or produces a match that is not a method group, check for the disposal  interface as described below.
+  - Perform overload resolution using the resulting method group and an empty argument list. If overload resolution results in no applicable methods, results in an ambiguity, or results in a single best method but that method is either static or not public, check for the disposal interface as described below.
+  - If the return type of the `DisposeAsync` method is not awaitable, an error is produced and no further steps are taken.
+  - The `finally` clause is expanded to the semantic equivalent of:
+  ```csharp
+    finally {
+        await e.DisposeAsync();
+    }
+    ```
+- Otherwise, if there is an implicit conversion from `E` to the `System.IAsyncDisposable` interface, then
+  - If `E` is a non-nullable value type then the `finally` clause is expanded to the semantic equivalent of:
+  ```csharp
+    finally {
+        await ((System.IAsyncDisposable)e).DisposeAsync();
+    }
+    ```
+  - Otherwise the `finally` clause is expanded to the semantic equivalent of:
+    ```csharp
+    finally {
+        System.IAsyncDisposable d = e as System.IAsyncDisposable;
+        if (d != null) await d.DisposeAsync();
+    }
+    ```
+    except that if `E` is a value type, or a type parameter instantiated to a value type, then the conversion of `e` to `System.IAsyncDisposable` shall not cause boxing to occur.
+- Otherwise, the `finally` clause is expanded to an empty block:
```

Comment:
If `E` is not a sealed type, should we generate the following, matching `foreach`?
```csharp
        System.IAsyncDisposable d = e as System.IAsyncDisposable;
        if (d != null) await d.DisposeAsync();
```

Reply:
I don't think so. I don't remember why we'd decided that, but the code has:
`// For async foreach, we don't do the runtime check`

```
             if ((!enumeratorType.IsSealed && !isAsync) ||
                    this.Conversions.ClassifyImplicitConversionFromType(enumeratorType,
                        isAsync ? this.Compilation.GetWellKnownType(WellKnownType.System_IAsyncDisposable) : this.Compilation.GetSpecialType(SpecialType.System_IDisposable),
                        ref useSiteInfo).IsImplicit)
                {
                    builder.NeedsDisposal = true;
                }
```